NS Agent auth with NS API

dynco-nym opened the pull request:
- NS Agent authenticates with key that was registered with NS API
- added flag to Agent to generate keypairs
- agent requests are signed by agent
- server-side requests are checked for authentication

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nymtech/nym/5127)
<!-- Reviewable:end -->

dynco-nym committed:
Agents authenticate with NSAPI

jstuczyn commented on nym-node-status-agent/src/cli/generate_keypair.rs:
```diff
@@ -0,0 +1,59 @@
+
+    let mut private_key_file = File::create(priv_key_path)?;
+    private_key_file.write_all(keypair.private_key().to_base58_string().as_bytes())?;
+
```
nitpicking: if unused, each private key should be zeroized

dynco-nym replied:
don't they implement [ZeroizeOnDrop](https://github.com/nymtech/nym/blob/develop/common/crypto/src/asymmetric/identity/mod.rs#L56), meaning they get zeroized a couple lines below?

jstuczyn replied:
the keypair itself, sure, but youre then allocating a base58 string (which you then convert to bytes). that one will not get zeroized.

as a side note: why not write key bytes directly instead?

jstuczyn commented on nym-node-status-agent/src/cli/mod.rs:
```diff
@@ -0,0 +1,74 @@
+
+        /// base58-encoded private key
+        #[arg(long, env = "NODE_STATUS_AGENT_AUTH_KEY")]
+        ns_api_auth_key: String,
```
I'm not sure if passing PRIVATE key explicitly in cli is a good idea because it will linger in history and args forever. I think it'd be safer (if possible) to pass path to the file since you're saving it anyway

dynco-nym replied:
it's not meant for passing via CLI: `clap` is just our convention because it handles loading from environment in a convenient way 

but for good measure, I removed the CLI option so now it's only loadable via env var

jstuczyn commented on nym-node-status-api/launch_node_status_api.sh:
```diff
@@ -7,6 +7,7 @@ export RUST_LOG=${RUST_LOG:-debug}
 export NYM_API_CLIENT_TIMEOUT=60
 export EXPLORER_CLIENT_TIMEOUT=60
 export NODE_STATUS_API_TESTRUN_REFRESH_INTERVAL=60
+export NODE_STATUS_API_AGENT_KEY_LIST="H4z8kx5Kkf5JMQHhxaW1MwYndjKCDHC7HsVhHTFfBZ4J,AKYZeKaJTzJrWtidk3BZz7wsm4GapaTY4GwQYVL43cmc"
```
uhm... I hope we didn't just dox ourselves by committing qa api keys : D

dynco-nym replied:
it's a keypair for testing only

jstuczyn commented on nym-node-status-api/src/cli/mod.rs:
```diff
@@ -56,19 +57,43 @@ pub(crate) struct Cli {
+#[derive(Debug, Clone)]
+struct KeyList(Vec<PublicKey>);
+
+fn parse_key_list(arg: &str) -> anyhow::Result<KeyList> {
```
clap can do it automatically by using `value_delimiter`

jstuczyn commented on nym-node-status-api/src/http/api/testruns.rs:
```diff
@@ -115,6 +133,23 @@ async fn submit_testrun(
     Ok(StatusCode::CREATED)
 }
 
+fn authenticate_agent(base58_pubkey: &str, state: &AppState) -> HttpResult<PublicKey> {
```
I'm sorry to say, but that's an AWFUL authentication. you're just sending your **public** key. anyone can do it since it's public. and even if it wasn't, it's susceptible to replay attacks.

jstuczyn replied:
ideally, each signed message would contain a unique nonce (bonus points if it's controlled by the server).

why not, instead, just go for a simpler `Bearer token` approach?

dynco-nym replied:
each request now contains
- payload 
  - containing unique timestamp (serving as nonce) for replay protection
- signature 
  - of the payload with agent's private key

jstuczyn commented on common/models/src/ns_api.rs:
```diff
@@ -1,7 +1,15 @@
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct SubmitResults {
+    pub message: String,
```
what's the content of this message? can we guarantee it has something unique per request? otherwise we're susceptible to replay attacks.

dynco-nym replied:
added `assigned_at_utc`: timestamp of when testrun was assigned. This is stored in the database for each testrun and checked upon submit

jstuczyn commented on common/models/src/ns_api.rs:
```diff
@@ -1,7 +1,15 @@
+pub struct SubmitResults {
+    pub message: String,
+    pub signature: Signature,
+    pub public_key: PublicKey,
```
shouldn't public key already be implicit given who has registered?

dynco-nym replied:
if agent didn't sent his own public key, server would have to check the database to find agent's public key BEFORE even verifying the request authenticity. That way unauthorized requests could spam-lock our database. 

This way at least server ensures
-  message is signed and 
- key is registered 

before doing further checks

jstuczyn commented on nym-node-status-api/src/http/api/testruns.rs:
```diff
@@ -28,10 +30,14 @@ pub(crate) fn routes() -> Router<AppState> {
+) -> HttpResult<Json<TestrunAssignment>> {
     // TODO dz log agent's network probe version
-    tracing::debug!("Agent requested testrun");
+    let agent_pubkey = authenticate_agent(&body, &state)?;
```
as mentioned in the method itself, this authentication is insufficient.

jstuczyn commented on nym-node-status-api/src/http/error.rs:
```diff
@@ -15,6 +15,13 @@ impl HttpError {
 
+    pub(crate) fn unauthorized() -> Self {
+        Self {
+            message: serde_json::json!({"message": "Make sure your public key si registered with NS API"}).to_string(),
```
```suggestion
            message: serde_json::json!({"message": "Make sure your public key is registered with NS API"}).to_string(),
```

jstuczyn commented on nym-node-status-agent/src/cli/run_probe.rs:
```diff
@@ -0,0 +1,119 @@
+    Ok(())
+}
+
+fn sign_message(key: PrivateKey, probe_outcome: String) -> SubmitResults {
```
key is not zeroized here

dynco-nym replied:
currently, it's used to sign a message, extract a public key, then it's dropped (zeroized). 

what would you like to do differently?

dynco-nym committed:
/submit with better auth
- also adjust agent run script to authenticate, even in parallel

dynco-nym committed:
/request better authentication
- moved agent API calls to Client struct

dynco-nym committed:
Replay protection

dynco-nym committed:
Fix testrun cleanup bug
- introduce a new column last_assigned which is different than
  created_at so that stale testruns get cleaned up based on
  last_assigned
- created_at is still useful for determining the "oldest" testrun
  to be picked up

dynco-nym committed:
Uniform request authentication

dynco-nym committed:
Suppress ts-rs serde warnings

jstuczyn commented on nym-node-status-agent/run.sh:
```diff
@@ -1,25 +1,52 @@
+# each key is required for a separate agent when running in parallel: their
+# public counterparts need to be registered with NS API
+private_keys=("BjyC9SsHAZUzPRkQR4sPTvVrp4GgaquTh5YfSJksvvWT"
+    "4RqSKydrEuyGF8Xtwyauvja62SAjqxFPYQzW2neZdkQD"
```
what are those keys? are those for some testnet? whatever they are, i dont think they should be hardcoded in a script...

dynco-nym replied:
exclusively for testing

jstuczyn commented on nym-node-status-agent/Cargo.toml:
```diff
@@ -16,12 +16,20 @@ readme.workspace = true
 [dependencies]
 anyhow = { workspace = true}
+bincode = { workspace = true }
+chrono = { workspace = true }
```
plz let's not use `chrono`, it caused us enough headache, if possible use `time` like we do everywhere else

dynco-nym replied:
as discussed in private, not a change for this PR but I agree.

dynco-nym committed:
Update cargo version

jstuczyn reviewed (approved):
🚢 it

dynco-nym committed:
All agents use the same key
- remove assigned_agent column
- remove logic which would stop agents with
  the same key to connect
- as a safety measure, add cap to total no. of agents
